### PR TITLE
[android] - harden orientation changes

### LINKED ISSUE #9127: [android] NullPointerException when rapidly changing screen orientation
**Platform:** Android
**Mapbox SDK version:** 5.1.0-beta.3

### Steps to trigger behavior

 Rapidly change screen orientation a couple of times

### Expected behavior

 App restores saved instance

### Actual behavior

App crashes

Here is crash insight:
`java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context com.mapbox.mapboxsdk.maps.MapView.getContext()' on a null object reference
                                                                                            at com.mapbox.mapboxsdk.maps.NativeMapView.getContext(NativeMapView.java:901)
                                                                                            at com.mapbox.mapboxsdk.maps.IconManager.loadIconForMarkerView(IconManager.java:78)
                                                                                            at com.mapbox.mapboxsdk.maps.AnnotationManager.prepareViewMarker(AnnotationManager.java:270)
                                                                                            at com.mapbox.mapboxsdk.maps.AnnotationManager.addMarker(AnnotationManager.java:229)
                                                                                            at com.mapbox.mapboxsdk.maps.MapboxMap.addMarker(MapboxMap.java:1123)
                                                                                            at mapboxdependencytest.MainActivity$1.onMapReady(MainActivity.java:35)
                                                                                            at com.mapbox.mapboxsdk.maps.MapView$MapCallback.onMapReady(MapView.java:964)
                                                                                            at com.mapbox.mapboxsdk.maps.MapView$MapCallback.access$1000(MapView.java:929)
                                                                                            at com.mapbox.mapboxsdk.maps.MapView$MapCallback$1.run(MapView.java:947)
                                                                                            at android.os.Handler.handleCallback(Handler.java:739)
                                                                                            at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                                            at android.os.Looper.loop(Looper.java:211)
                                                                                            at android.app.ActivityThread.main(ActivityThread.java:5321)
                                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                                            at java.lang.reflect.Method.invoke(Method.java:372)
                                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1016)
                                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:811)`

and code which creates it:
`public class MainActivity extends AppCompatActivity {

    private MapView mapView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Mapbox.getInstance(this, "...");

        setContentView(R.layout.activity_main);

        mapView = (MapView) findViewById(R.id.map_view);

        mapView.onCreate(savedInstanceState);

        mapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(final MapboxMap mapboxMap) {
                mapboxMap.addMarker(new MarkerViewOptions().position(new LatLng(51,16))); //exception is thrown on this line
                mapboxMap.getMarkerViewManager().setOnMarkerViewClickListener(new MapboxMap.OnMarkerViewClickListener() {
                    @Override
                    public boolean onMarkerClick(@NonNull Marker marker, @NonNull View view, @NonNull MapboxMap.MarkerViewAdapter adapter) {
                        Toast.makeText(MainActivity.this, "Marker click", Toast.LENGTH_SHORT).show();
                        return true;
                    }
                });
            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mapView.onSaveInstanceState(outState);
    }

    @Override
    public void onStart() {
        super.onStart();
        mapView.onStart();
    }

    @Override
    public void onStop() {
        super.onStop();

        mapView.onStop();
    }

    @Override
    public void onResume() {
        super.onResume();
        mapView.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        mapView.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }
}`

### PR DESCRIPTION
**Pull request by tobrun**
closes #9127

### CODE REVIEW

**Comment by LukasPaczos** [comment]
I applied your changes but unfortunatelly the issue [#9127](https://github.com/mapbox/mapbox-gl-native/issues/9127) is still present.

`java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context com.mapbox.mapboxsdk.maps.MapView.getContext()' on a null object reference
                                                                                                     at com.mapbox.mapboxsdk.maps.NativeMapView.getContext(NativeMapView.java:901)
                                                                                                     at com.mapbox.mapboxsdk.maps.IconManager.loadIconForMarkerView(IconManager.java:78)
                                                                                                     at com.mapbox.mapboxsdk.maps.AnnotationManager.prepareViewMarker(AnnotationManager.java:270)
                                                                                                     at com.mapbox.mapboxsdk.maps.AnnotationManager.addMarker(AnnotationManager.java:229)
                                                                                                     at com.mapbox.mapboxsdk.maps.MapboxMap.addMarker(MapboxMap.java:1123)`

**Comment by tobrun** [comment]
@LukasPaczos Thank you for testing this so quickly, I wasn't able to reproduce it with a SimpleActivity. Let me try running the sample code you provided.

**Comment by LukasPaczos** [comment]
I've done some logging and that's what i got:

Map: new nativeMapView = 917213383
Map: onStart()
Map: onResume()
Map: getMapAsync callback
Map: onPause()
Map: onStop()
Map: onDestory()
Map: NativeMapView.destroy(), nativeMapView = 917213383
Map: new nativeMapView = 892499314
Map: onStart()
Map: onResume()
Map: onPause()
Map: onStop()
Map: onDestory()
Map: NativeMapView.destroy(), nativeMapView = 892499314
Map: new nativeMapView = 422282973
Map: onStart()
Map: onResume()
Map: getMapAsync callback
Map: onPause()
Map: onStop()
Map: onDestory()
Map: NativeMapView.destroy(), nativeMapView = 422282973
Map: new nativeMapView = 173994215
Map: onStart()
Map: onResume()
Map: getMapAsync callback
Map: getMapAsync callback
Map: onPause()
Map: onStop()
Map: onDestory()
Map: NativeMapView.destroy(), nativeMapView = 173994215
Map: new nativeMapView = 62699327
Map: onStart()
Map: onResume()
Map: loadIconForMarkerView - nativeMapView = 892499314

Those numbers are hashcodes of nativeMapViews and as you can see, the reference of nativeMapView in the moment of crash should be long gone and overwritten as it was killed 2 cycles prior.

**Comment by tobrun** [comment]
@LukasPaczos would you be able to take e48fe34 for a spin?

**Comment by LukasPaczos** [comment]
@tobrun sure, first thing in the morning

**Comment by LukasPaczos** [comment]
@tobrun [e48fe34](https://github.com/mapbox/mapbox-gl-native/commit/e48fe34be0662b8e46247c1e49e5d24a0997564d) does seem to patch things up!

**Comment by tobrun** [comment]
@LukasPaczos thank you for testing, 
I will ticket out the double OnMapReady invoke, this is indeed a separate issue.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/IconManager.java [hunk 1]
```diff
@@ -40,56 +41,60 @@
   }
 
   Icon loadIconForMarker(Marker marker) {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/IconManager.java]: This method is really long. What about extracting some blocks of code into `private` methods?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/IconManager.java [hunk 2]
```diff
@@ -126,25 +131,28 @@ void reloadIcons() {
...
       }
-    }
 
-    // this seems to be a costly operation according to the profiler so I'm trying to save some calls
-    Marker previousMarker = marker.getId() != -1 ? (Marker) mapboxMap.getAnnotation(marker.getId()) : null;
-    if (previousMarker == null || previousMarker.getIcon() == null || previousMarker.getIcon() != marker.getIcon()) {
-      marker.setTopOffsetPixels(getTopOffsetPixelsForIcon(icon));
+      // this seems to be a costly operation according to the profiler so I'm trying to save some calls
+      Marker previousMarker = marker.getId() != -1 ? (Marker) mapboxMap.getAnnotation(marker.getId()) : null;
+      if (previousMarker == null || previousMarker.getIcon() == null || previousMarker.getIcon() != marker.getIcon()) {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/IconManager.java]: Expression hard to read. What about breaking it down introducing explaining variables?

File: platform/android/build.gradle [hunk 3]
```diff
@@ -3,7 +3,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.1'
+        classpath 'com.android.tools.build:gradle:2.4.0-alpha7'
```
Comment (Guardiola31337) [inline platform/android/build.gradle]: Are we going to bump gradle tools into an alpha version?

**Comment by tobrun** [comment]
@Guardiola31337 review comments have been addressed

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Icon.java [hunk 4]
```diff
@@ -33,9 +36,32 @@ public String getId() {
    * @return The {@link Bitmap} being used for the {@link Icon}.
    */
   public Bitmap getBitmap() {
+    if (mBitmap.getConfig() != Bitmap.Config.ARGB_8888) {
+      mBitmap = mBitmap.copy(Bitmap.Config.ARGB_8888, false);
+    }
     return mBitmap;
   }
 
+  public float getScale() {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Icon.java]: We should provide javadoc on all `public` methods.

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Icon.java [hunk 5]
```diff
@@ -33,9 +36,32 @@ public String getId() {
...
+      throw new IllegalStateException("Required to set a Icon before calling getScale");
+    }
+    float density = mBitmap.getDensity();
+    if (density == Bitmap.DENSITY_NONE) {
+      density = DisplayMetrics.DENSITY_DEFAULT;
+    }
+    return density / DisplayMetrics.DENSITY_DEFAULT;
+  }
+
+  public byte[] toBytes(){
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Icon.java]: javadoc

**Pull request merged**